File: src/phorge/api/endpoints/firewall.py
"""Firewall rule API endpoints."""

from __future__ import annotations

from typing import Any

from phorge.api.client import ForgeClient
from phorge.api.models import FirewallRule


class FirewallAPI:
    def __init__(self, client: ForgeClient) -> None:
        self._client = client

    async def list(self, server_id: int) -> list[FirewallRule]:
        data = await self._client.get(f"/servers/{server_id}/firewall-rules")
        return [FirewallRule(**r) for r in data.get("rules", [])]

    async def get(self, server_id: int, rule_id: int) -> FirewallRule:
        data = await self._client.get(f"/servers/{server_id}/firewall-rules/{rule_id}")
        return FirewallRule(**data["rule"])

    async def create(
        self,
        server_id: int,
        *,
        name: str,
        port: int | str,
        ip_address: str | None = None,
        type: str = "allow",
    ) -> FirewallRule:
        payload: dict[str, Any] = {
            "name": name,
            "port": port,
            "type": type,
        }
        if ip_address:
            payload["ip_address"] = ip_address
        data = await self._client.post(
            f"/servers/{server_id}/firewall-rules", json=payload
        )
        return FirewallRule(**data["rule"])

    async def delete(self, server_id: int, rule_id: int) -> None:
        await self._client.delete(f"/servers/{server_id}/firewall-rules/{rule_id}")

File: internal/forge/firewall.go
package forge

import (
	"context"
	"net/http"
)

// FirewallCreateOpts contains the options for creating a firewall rule.
type FirewallCreateOpts struct {
	Name      string `json:"name"`
	Port      string `json:"port,omitempty"`
	IPAddress string `json:"ip_address,omitempty"` // optional
	Type      string `json:"type"`                 // default "allow"
}

// List returns all firewall rules on a server.
func (s *FirewallService) List(ctx context.Context, serverID int64) ([]FirewallRule, error) {
	return listResources[FirewallRule](s.client, ctx, s.client.orgPath("/servers/%d/firewall-rules", serverID))
}

// Get returns a single firewall rule by ID.
func (s *FirewallService) Get(ctx context.Context, serverID, ruleID int64) (*FirewallRule, error) {
	return getResource[FirewallRule](s.client, ctx, s.client.orgPath("/servers/%d/firewall-rules/%d", serverID, ruleID))
}

// Create creates a new firewall rule on a server.
func (s *FirewallService) Create(ctx context.Context, serverID int64, opts FirewallCreateOpts) (*FirewallRule, error) {
	return createResource[FirewallRule](s.client, ctx, s.client.orgPath("/servers/%d/firewall-rules", serverID), opts)
}

// Delete removes a firewall rule from a server.
func (s *FirewallService) Delete(ctx context.Context, serverID, ruleID int64) error {
	return s.client.do(ctx, http.MethodDelete, s.client.orgPath("/servers/%d/firewall-rules/%d", serverID, ruleID), nil, nil)
}
